### PR TITLE
Add image tags for kubelet

### PR DESCRIPTION
**Pull request by mfpierre**
### What does this PR do?

Add image tags for the kubelet tagger

### Motivation

Have image tags if running in kubelet-only mode

### Additional Notes

If kubernetes is running in docker runtime image tags should be both collected from doker & kubelet tagger but deduped in the store as they should be equivalent

### CODE REVIEW

**Review by hkaj — DISMISSED** [review]

File: pkg/tagger/collectors/kubelet_extract.go [hunk 1]
```diff
@@ -99,6 +100,25 @@ func (c *KubeletCollector) parsePods(pods []*kubelet.Pod) ([]*TagInfo, error) {
 		// container tags
 		for _, container := range pod.Status.Containers {
 			lowC := append(low, fmt.Sprintf("kube_container_name:%s", container.Name))
+			// check image tag in spec
+			for _, containerSpec := range pod.Spec.Containers {
+				if containerSpec.Name == container.Name {
+					imageName, shortImage, imageTag, err := docker.SplitImageName(containerSpec.Image)
+					if err != nil {
+						log.Debugf("Error splitting %s: %s", containerSpec.Image, err)
```
Comment (JulienBalestra) [inline pkg/tagger/collectors/kubelet_extract.go]: We should avoid the keyword `Error` and instead use words like `Cannot`. Especially when the severity is different than error.

File: pkg/tagger/tagstore.go [hunk 2]
```diff
@@ -131,31 +138,65 @@ func (e *entityTags) get(highCard bool) ([]string, []string) {
...
+
+	lowCardTags := []string{}
+	highCardTags := []string{}
+	for i := 0; i < len(tagSortList); i++ {
+		insert := true
+		for j := 0; j < len(tagSortList); j++ {
+			// if we find a duplicate tag with higher priority we do not insert the tag
+			if i != j && tagSortList[i].tagName == tagSortList[j].tagName &&
+				collectors.CollectorPriorities[tagSortList[i].source] < collectors.CollectorPriorities[tagSortList[j].source] {
+				insert = false
```
Comment (xvello) [inline pkg/tagger/tagstore.go]: we could `break` too, no need to compare other items

File: pkg/tagger/tagstore.go [hunk 3]
```diff
@@ -131,31 +138,65 @@ func (e *entityTags) get(highCard bool) ([]string, []string) {
...
+
+	tags := append(lowCardTags, highCardTags...)
 
 	// Write cache
 	e.RUnlock()
 	e.Lock()
 	e.cacheValid = true
 	e.cachedSource = sources
 	e.cachedAll = tags
-	e.cachedLow = e.cachedAll[:lowCardCount]
```
Comment (xvello) [inline pkg/tagger/tagstore.go]: nitpick, but this allows the `cachedLow` to be a slice of the same backing array `cachedAll` uses, and reduces the memory usage of the tagstore cache. You could keep it easily by setting `e.cachedLow = e.cachedAll[len(lowCardTags)]`, then `lowCardTags` will be GCed

File: pkg/tagger/tagstore_test.go [hunk 4]
```diff
@@ -179,11 +179,54 @@ func TestGetEntityTags(t *testing.T) {
 	etags.cacheValid = false
 	tags, sources = etags.get(true)
 	assert.Len(t, tags, 4)
-	assert.Contains(t, tags, "low1", "low2", "high1", "high2")
+	assert.ElementsMatch(t, tags, []string{"low1", "low2", "high1", "high2"})
```
Comment (xvello) [inline pkg/tagger/tagstore_test.go]: 🙇

File: releasenotes/notes/kubelet-image-tags-b74a4dc514a33a58.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    The kubelet will now collect image tags from the pod list
```
Comment (xvello) [inline releasenotes/notes/kubelet-image-tags-b74a4dc514a33a58.yaml]: What about "The agent will now collect container image tags when running in kubernetes-only mode" ?

File: pkg/tagger/tagstore.go [hunk 6]
```diff
@@ -131,31 +138,65 @@ func (e *entityTags) get(highCard bool) ([]string, []string) {
 	}
 
 	// Cache miss
-	var arrays [][]string
 	var sources []string
-	lowCardCount := 0
+	var tagSortList []tagSort
```
Comment (xvello) [inline pkg/tagger/tagstore.go]: To make lookup faster, what do you think about making `tagSortList` a map, with `tagName` as key, so the main case (tags are not duplicated) does not incur a full list traversal?

**Review by xvello — CHANGES_REQUESTED** [review]
Logic makes the job without impacting too much the tagStore 👍 . We could optimise it later by changing the `tagInfo` structure, to avoid `strings.Split`-ing everytime.

Some nits on readability and performance

File: pkg/tagger/tagstore.go (hunk 6 above)
Comment (mfpierre) [reply pkg/tagger/tagstore.go]: That was my first approach but I had a hard time doing it, but I was able to iterate on this first draft, should be much cleaner now

Comment (mfpierre) [reply pkg/tagger/tagstore.go]: I'm mixed about the results https://github.com/DataDog/datadog-agent/pull/1383/commits/edd302dd7deae03399db40d586c25829f3718811

Might be a little more optimised but it's adding another loop level, WDYT?

File: pkg/tagger/tagstore.go [hunk 7]
```diff
@@ -131,31 +137,64 @@ func (e *entityTags) get(highCard bool) ([]string, []string) {
...
-	return tags[:lowCardCount], sources
+	return lowCardTags, sources
+}
+
+func insertWithPriority(tagSortList map[string][]tagSort, tags []string, source string, isHighCard bool) {
+	for _, t := range tags {
+		tagName := strings.Split(t, ":")[0]
+		tagSortList[tagName] = append(tagSortList[tagName], tagSort{
+			tag:        t,
+			priority:   collectors.CollectorPriorities[source],
```
Comment (xvello) [inline pkg/tagger/tagstore.go]: The priority lookup could be moved before the loop, with a big fat warnf if the lookup fails

File: pkg/tagger/tagstore.go [hunk 8]
```diff
@@ -131,31 +137,64 @@ func (e *entityTags) get(highCard bool) ([]string, []string) {
 	}
 
 	// Cache miss
-	var arrays [][]string
 	var sources []string
-	lowCardCount := 0
+	tagSortList := make(map[string][]tagSort)
```
Comment (xvello) [inline pkg/tagger/tagstore.go]: we're not sorting the tags, but de-duplicating them, could we use a more explicit name for both types?

**Review by xvello — DISMISSED** [review]
LGTM! Two last nits

**Review by xvello — APPROVED** [review]
LGTM 👍

**Pull request merged**